src/nxt_cgroup.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_process.h>
+#include <nxt_fs.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <limits.h>
+
+
+static nxt_int_t nxt_cgroup_create(const nxt_task_t *task,
+    nxt_process_t *process);
+
+
+static nxt_int_t
+nxt_cgroup_create(const nxt_task_t *task, nxt_process_t *process)
+{
+    char  cgpath[PATH_MAX];
+
+    if (task->thread->runtime->type != NXT_PROCESS_MAIN
+        || nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.name == NULL)
+    {
+        return NXT_OK;
+    }
+
+    snprintf(cgpath, sizeof(cgpath), NXT_CGROUP_ROOT "/%s",
+             process->isolation.cgroup.name);
+    nxt_fs_mkdir_all((const u_char *)cgpath, 0777);
+
+    return NXT_OK;
+}
+
+
+nxt_int_t
+nxt_cgroup_proc_add(const nxt_task_t *task, nxt_process_t *process)
+{
+    FILE  *fp;
+    char  cgprocs[PATH_MAX];
+
+    nxt_cgroup_create(task, process);
+
+    if (nxt_process_type(process) != NXT_PROCESS_PROTOTYPE
+        || process->isolation.cgroup.name == NULL)
+    {
+        return NXT_OK;
+    }
+
+    snprintf(cgprocs, sizeof(cgprocs), NXT_CGROUP_ROOT "/%s/cgroup.procs",
+             process->isolation.cgroup.name);
+    fp = fopen(cgprocs, "we");
+    if (fp == NULL) {
+        return NXT_ERROR;
+    }
+
+    fprintf(fp, "%d\n", process->pid);
```

Comment:
it's advisable to always check for write and close errors and avoid buffering APIs when doing sysfs-like operations in the kernel. In the case of cgroups, the write(2) can fail in [many ways](https://github.com/torvalds/linux/blob/master/kernel/cgroup/cgroup.c#L3906:L3957) (EINVAL, EPERM, etc). Even if you chose to continue, if the user asked for the cgroup isolation and Unit failed to do so, we should not silently ignore it -- IMHO.

Reply:
Valid point on the buffering. A
```C
setvbuf(fp, NULL, _IONBF, 0);
```
after the fopen(3) should fix that. That way I can keep the fprintf(3) and don't need to faff around with temporary variables and converting int's to strings and all that jazz....

Reply:
The biggest issue here is not checking the error... The `fclose(3)` must also have its error checked because the last `write(2)` error can be returned in the close() call...

Other than that, using `fprintf(3)` is not the idiomatic way of writing to files in the NGINX Unit code base. It's only used 2x in the entire codebase and for writing to stderr... and only in the nxt_unit.c (the language lib) because it can't link the nxt_* core libs.

Below is a more idiomatic way:

```c
u_char      *end;
ssize_t     length;
nxt_file_t  file;
u_char      pid[NXT_INT64_T_LEN + nxt_length("\n")];

// open file with nxt_file_open()
// ...

end = nxt_sprintf(pid, pid + sizeof(pid), "%PI%n", process->pid);
length = end - pid;

if (nxt_file_write(&file, pid, length, 0) != length) {
    return NXT_ERROR;
}

nxt_file_close(task, &file);

return NXT_OK
```
This code was adapted from the [nxt_runtime_pid_file_create](https://github.com/nginx/unit/blob/master/src/nxt_runtime.c#L1361) function.

Reply:
Since buffering has been disabled, *fclose(3)* is only allowed to fail if the file descriptor is not valid (`EBADF`).  Checking errors only from *frpintf(3)* should be fine.

I think *fprintf(3)* is an elegant solution, since it avoids having to create a temporary string to do the formatting.

Reply:
> Below is a more idiomatic way:

I have to admit, I'm not a huge fan.
 
> ```c
> u_char      *end;
> ssize_t     length;
> nxt_file_t  file;
> u_char      pid[NXT_INT64_T_LEN + nxt_length("\n")];
>```
Non NUL terminated string? Sure, it's unlikely pids will get to that size and we're dealing with it in terms of length, but I think I'd just make that '+ 2'...

>```C
> // open file with nxt_file_open()
> // ...
> 
> end = nxt_sprintf(pid, pid + sizeof(pid), "%PI%n", process->pid);
> length = end - pid;
>```
A little cumbersome... hmm, so you have to explicitly ask for it to be NUL terminated with _%Z_...
>```C
> if (nxt_file_write(&file, pid, length, 0) != length) {
>     return NXT_ERROR;
> }
>```
File descriptor leak.
>```C
> nxt_file_close(task, &file);
> 
> return NXT_OK
> ```

Reply:
> Since buffering has been disabled, fclose(3) is only allowed to fail if the file descriptor is not valid (EBADF)


@alejandro-colomar in general, this is not the case, have a look in the `NOTES` section of the `fclose(3)` man page. It also reports the `close(2)` errors.

The `setvbuf(3)` only controls the libc buffer but filesystem drivers also make use of internal caches, then the `close(2)` system call can still fail if the previous `write(2)` cached the data but failed to persist to underlying storage, as stated here:
https://man7.org/linux/man-pages/man2/close.2.html#:~:text=Dealing%20with%20error%20returns%20from,releases%20the%20open%20file%20description.

There's nothing much you can do in this case other than log the error and continue, but this can be very important while troubleshooting issues.

But if you mean that "*in the case of the cgroups fs*, the fclose(3) is only allowed to fail if the file descriptor is not valid ..." then ok, I really don't know, but the syscall close(2) API returns an error so it's allowed to fail in the future, then in this case Unit will silently ignore it.

Reply:
@ac000 Sorry for the buggy snippet but the intention was to show the alternative API.
You're right, in the case of kernel APIs, strings are commonly NUL terminated.

Reply:
> I think fprintf(3) is an elegant solution, since it avoids having to create a temporary string to do the formatting.

Fair enough, just pointing out that this `fprintf(3)` stands out from the rest of the code, but if the errors are checked, all good.

Reply:
> > Since buffering has been disabled, fclose(3) is only allowed to fail if the file descriptor is not valid (EBADF)
> 
> @alejandro-colomar in general, this is not the case, have a look in the `NOTES` section of the `fclose(3)` man page. It also reports the `close(2)` errors.
> 
> The `setvbuf(3)` only controls the libc buffer but filesystem drivers also make use of internal caches, then the `close(2)` system call can still fail if the previous `write(2)` cached the data but failed to persist to underlying storage, as stated here: https://man7.org/linux/man-pages/man2/close.2.html#:~:text=Dealing%20with%20error%20returns%20from,releases%20the%20open%20file%20description.
> 
> There's nothing much you can do in this case other than log the error and continue, but this can be very important while troubleshooting issues.
> 
> But if you mean that "_in the case of the cgroups fs_, the fclose(3) is only allowed to fail if the file descriptor is not valid ..." then ok, I really don't know, but the syscall close(2) API returns an error so it's allowed to fail in the future, then in this case Unit will silently ignore it.

Yeah, I expect the cgroups fs to be safe from any caching, since there's no real fs.  We could still check *fclose(2)* errors, in case the kernel decides to fail at *close(2)* for some unicorn reason in the future.  But yeah, none of `EINTR` `EIO` `ENOSPC` nor `EDQUOT` make sense to me in this pseudo-fs.